### PR TITLE
Move Kernels.cu to Alpaka

### PR DESCRIPTION
**Pull request by GNiendorf**
Moves the kernels found in Kernels.cu to Kernels.cuh and over to alpaka. I also deleted Kernels.cu since there was no code left in the file after the change. This PR also removes the unnecessary Constants.h file.

One note, I had to change the thread size from (1, 32, 32) to (1, 16, 16) for the "removeDupPixelQuintupletsInGPUFromMap" kernel since the original thread size resulted in a runtime error relating to invalid configuration. The maximum threads per block is 1024, so I'm not sure why (1, 32, 32) throws an error, but (1, 16, 16) is faster anyway as you can see from the timing. This does not throw an error in the CUDA version, so this error is specific to something in Alpaka. 

edit: See this issue I made in the Alpaka github: https://github.com/alpaka-group/alpaka/issues/1943

Kernels with altered thread sizes (1,32,32)->(1,16,16) motivated by the improvement above:
1. removeDupPixelQuintupletsInGPUFromMap
2. CheckHitspLS
3. removeDupPixelTripletsInGPUFromMap
4. removeDupQuintupletsInGPUAfterBuild

Final Timing for This PR
Screenshot 2023-04-15 at 8 09 03 PM

This PR Timing (Outdated)
Screenshot 2023-04-11 at 1 51 06 PM

Current Alpaka_Move Timing
Screenshot 2023-04-11 at 2 28 33 PM

Validation Plots (Outdated): [Here](https://www.classe.cornell.edu/~gsn27/www/www/PR270/mywork_9dc308-PU200/summary/)

Comparison to Master (Outdated): [Here](https://www.classe.cornell.edu/~gsn27/www/www/PR270/mywork_398270-PU200_9dc308-PU200/summary/)

### CODE REVIEW

File: code/core/constants.h
Comment (VourMa) [inline code/core/constants.h:4]: Initially I was confused about the inclusion of this file to other files but now I understand that this definition is needed. However, this is not the correct pT cut... That prompts me to ask a question larger than this PR but possibly connected to the alpaka migration:
Are most of the files in the `code/core` directory supposed to be a CPU version of our code? If so, isn't the CPU code outdated and also wrong (as details like the one that prompted this comment seem to imply)? It's also not used, as far as I know.
Should we remove it, given the CPU version coming automatically with the move to the alpaka framework? Maybe @sgnoohc is the best person to ask. We should possibly make an issue to discuss and understand if there are any action items here for a future PR.

**Review by VourMa — COMMENTED** [review]
I only had minor comments for the actual PR, while the most important comments focus on two more general issues:
1. This PR includes a good amount of clean-up, i.e. removal of functions that are not used anywhere. Before removing them, I would like to understand their purpose and how they fell out of use, just to make sure that we don't delete something that is there for a reason. @tresreid could maybe take a look at the comments of this PR and help with understanding whether we should really nuke those unused functions?
2. Updates in this PR bring up the issue of outdated code in the `code/core` directory, so I would like to take this opportunity to start a discussion to understand the situation. Of course, this is not as an action item for this PR but will possibly go into a separate issue.

File: SDL/Kernels.cu
Comment (GNiendorf) [reply SDL/Kernels.cu:32]: Yup, they aren't used anywhere

File: SDL/Constants.h
Comment (GNiendorf) [reply SDL/Constants.h:80]: Correct, none of these constants are used anywhere (except one case in the old cpu code that is in code/core which is compiled but never used I think). You can check for any instances of SDL::CPU::Constants or whatever the tag here is.

File: SDL/Kernels.cuh
Comment (GNiendorf) [reply SDL/Kernels.cuh:13]: Just to harmonize with the other files and avoid any potential conflicts.

File: SDL/Kernels.cu
Comment (GNiendorf) [reply SDL/Kernels.cu:53]: Correct, not used anywhere.

File: SDL/Kernels.cu
Comment (GNiendorf) [reply SDL/Kernels.cu:372]: Yup, not used anywhere.

File: code/core/constants.h
Comment (GNiendorf) [reply code/core/constants.h:4]: From the looks of it, I think there was a mistake when the cpu code was moved out of SDL? If you look, there are duplicate files in the cpu directory and code/core. So there are a lot of files in code/core that are just the old cpu files. My plan was to remove all of the old cpu code in a subsequent PR.

**Comment by tresreid** [comment]
Also, I think the reason constants.h was .h and not .cuh was because it is used in code/core/write_sdk_ntuple.h. I'm not sure if a .cuh file will work there or not.
It is needed in that file in case we switch around the half vs full precision. I think changing the data types can cause issues if they don't match up. Are you running alpaka with half precision? If not, perhaps that is why it seems to run ok now? What is your memory usage with alpaka?

**Comment by tresreid** [comment]
> Also, I think the reason constants.h was .h and not .cuh was because it is used in code/core/write_sdk_ntuple.h. I'm not sure if a .cuh file will work there or not. It is needed in that file in case we switch around the half vs full precision. I think changing the data types can cause issues if they don't match up. Are you running alpaka with half precision? If not, perhaps that is why it seems to run ok now? What is your memory usage with alpaka?

Actually browsing the code, I don't see the half precision switch in the write_sdl_ntuple file. Which is sort of strange. I don't remember it being removed. But if that is the case than constants.cu -> .cuh seems reasonable.
And I guess you aren't removing Kernels.cu just moving everything to kernels.cuh. which seems fine. though perhaps renaming it would still be good. 
Otherwise, I think this looks good.

**Comment by VourMa** [comment]
Thanks for going through the PR and the comments, @tresreid!

> browsing the comments, the "scoreX" functions can probably be removed. These were historically used to get the r-z chi2 values which were used for duplicate removal. But those are now implemented in the runtime of building the objects and the scores are saved directly there.
The checkX and rmXtomemory should all be kept as they are still used to look for duplicates between objects and to mark objects as duplicates.

Ok, that's exactly what we do here and the main of my question. So the PR is good to go once Gavin finalizes the few minor comments left.

> A lot of the functions within code/core are used for the CPU part and should be purged but I think there was an issue doing this before. I think the issue was that a lot of the validation and writing to the Ntuple stuff uses some of these functions. @sgnoohc would know best since the CPU side of things were all his domain.

Sure. I guess we will hit that wall when Gavin makes the relevant PR and we can involve Philip in the loop then, so that everything is done properly.

> It is my preference that constants.cu should be kept. I think this should be our config file where we define all of our constants and cuts. So if we need to change something, we change it there instead of having to dig through 100000 lines for code to figure out what needs to be changed. Take for example the PT cut that we are still apparently having issues with. Since this is implemented throughout the code, when we wanted to change it we had to find all the instances of it being used. If this was set in one easy to find place we wouldn't still be having this issue. Same for all other cuts. And in the future we can feed this into some ML implementation (like TMVA) to help try to optimize these cuts. I think it is worth the clarity to keep.

I totally agree and this is the direction we are moving with this PR, right? Our goal would be to have only `Constants.cuh` and `Constants.cu` which will include all of the constants and "configuration variables" throughout the code.

> Kernel.cu I would prefer if we kept since many of those functions are used for duplicate removal and are used in several places for different objects. For example, I believe the checkPLs function is used for pT5, pT3 dup rm and cross cleaning. I think this is the best place for it rather than within the larger pixelobjects file(s). Maybe rename it to DuplicateRemoval.cu or something that clarifies the purpose. I don't think having the extra file hurts us in any real way here.

Unless I have misunderstood, the plan was to anyways keep the `Kernels` files. But all the functions had to be moved from `Kernels.cu` to the `Kernels.cuh` file due to [this](https://github.com/SegmentLinking/TrackLooper/pull/243#discussion_r1124803787).
I am fine with renaming as long as the name is descriptive, so I leave it up to @GNiendorf to decide.

**Comment by VourMa** [comment]
> Also, I think the reason constants.h was .h and not .cuh was because it is used in code/core/write_sdk_ntuple.h. I'm not sure if a .cuh file will work there or not.

But the validation plots, which run the NTuple, worked fine, so it should be fine, right?

> It is needed in that file in case we switch around the half vs full precision. I think changing the data types can cause issues if they don't match up. Are you running alpaka with half precision? If not, perhaps that is why it seems to run ok now? What is your memory usage with alpaka?

My understanding is that we are still running with half precision. Alpaka does not support it but it works since we are only compiling for the "cuda" backend up till now (@GNiendorf please correct me if my understanding is not right). In the long run, we would probably have to either drop it or check with the developers if it's going to be supported in some way. I would leave that as a final step in the migration but of course Gavin is the best person to decide that.
Memory usage should be checked at some point - I guess this can happen once all of the kernels have been migrated and before we migrate the memory transfers?

**Comment by tresreid** [comment]
> 

Yeah, This is something I would have to look into but if everything works fine as is then i guess it isn't an issue.

**Comment by tresreid** [comment]
> Thanks for going through the PR and the comments, @tresreid!
> 
> > browsing the comments, the "scoreX" functions can probably be removed. These were historically used to get the r-z chi2 values which were used for duplicate removal. But those are now implemented in the runtime of building the objects and the scores are saved directly there.
> > The checkX and rmXtomemory should all be kept as they are still used to look for duplicates between objects and to mark objects as duplicates.
> 
> Ok, that's exactly what we do here and the main of my question. So the PR is good to go once Gavin finalizes the few minor comments left.
> 
> > A lot of the functions within code/core are used for the CPU part and should be purged but I think there was an issue doing this before. I think the issue was that a lot of the validation and writing to the Ntuple stuff uses some of these functions. @sgnoohc would know best since the CPU side of things were all his domain.
> 
> Sure. I guess we will hit that wall when Gavin makes the relevant PR and we can involve Philip in the loop then, so that everything is done properly.
> 
> > It is my preference that constants.cu should be kept. I think this should be our config file where we define all of our constants and cuts. So if we need to change something, we change it there instead of having to dig through 100000 lines for code to figure out what needs to be changed. Take for example the PT cut that we are still apparently having issues with. Since this is implemented throughout the code, when we wanted to change it we had to find all the instances of it being used. If this was set in one easy to find place we wouldn't still be having this issue. Same for all other cuts. And in the future we can feed this into some ML implementation (like TMVA) to help try to optimize these cuts. I think it is worth the clarity to keep.
> 
> I totally agree and this is the direction we are moving with this PR, right? Our goal would be to have only `Constants.cuh` and `Constants.cu` which will include all of the constants and "configuration variables" throughout the code.
> 
> > Kernel.cu I would prefer if we kept since many of those functions are used for duplicate removal and are used in several places for different objects. For example, I believe the checkPLs function is used for pT5, pT3 dup rm and cross cleaning. I think this is the best place for it rather than within the larger pixelobjects file(s). Maybe rename it to DuplicateRemoval.cu or something that clarifies the purpose. I don't think having the extra file hurts us in any real way here.
> 
> Unless I have misunderstood, the plan was to anyways keep the `Kernels` files. But all the functions had to be moved from `Kernels.cu` to the `Kernels.cuh` file due to [this](https://github.com/SegmentLinking/TrackLooper/pull/243#discussion_r1124803787). I am fine with renaming as long as the name is descriptive, so I leave it up to @GNiendorf to decide.

Yeah ok. I initially misunderstood from the comments what was happening when you were removing those files. But understanding now that the functions are being moved to header files, I think everything is fine.

**Comment by GNiendorf** [comment]
Timing
Screenshot 2023-04-15 at 8 09 03 PM

Comparison Plots: [Here](https://www.classe.cornell.edu/~gsn27/www/www/PR270/recent_comp/summary/)

**Review by VourMa — APPROVED** [review]
Worth highlighting that the alpaka bug(?) with the number of threads prompted us to check lower number of threads for the duplicate removal kernels and that led to a 10% timing reduction.

**Pull request merged**
